Format URL encoded forms

gm-vm opened the pull request:
## :camera: Screenshots
|Current|Formatted|
|-|-|
|![current](https://user-images.githubusercontent.com/54670603/79045408-3db55580-7c0b-11ea-8724-2e29d094fc8c.png)|![formatted](https://user-images.githubusercontent.com/54670603/79045412-40b04600-7c0b-11ea-97c0-f608407f54df.png)|

## :page_facing_up: Context
Reading encoded forms is nearly impossible, especially when dealing values have symbols in them. Unlike for XML and JSON requests, this does not simply pretty print the body and that may be undesired.

gm-vm commented:
I thought about connecting the existing button, but I had some doubts and, honestly, I don't know why I didn't mention them.

What I've just come up with is the following: I'd add a new button that allows to switch between two display modes: "raw" and "formatted". This button could be either the existing button or a new one, not sure what's better. I simply don't want to add request-specific parameters to `HttpTransaction` and this is the only way to do that. I'd also include responses, just for consistency.

But how should the share button behave? Would it follow the current display mode? Should it include both the formatted request and the raw request? Just the one currently displayed?

gm-vm commented on library/src/main/java/com/chuckerteam/chucker/internal/support/FormatUtils.kt:
```diff
@@ -94,6 +96,19 @@ internal object FormatUtils {
         }
     }
 
+    fun formatUrlEncodedForm(form: String): String {
```
Sure, I'll also rewrite the method not to make any assumption on the input.

cortinico commented:
> This button could be either the existing button or a new one, not sure what's better. I simply don't want to add request-specific parameters to `HttpTransaction` and this is the only way to do that. I'd also include responses, just for consistency.

Agree. The stored information should be unchanged. What we're addressing here is just on the presentation layer.

If possible I would hook to the same button we already have on the URL side. The semantic is pretty clear as it's an encoded/decoded toggle.

gm-vm commented:
> Agree. The stored information should be unchanged. What we're addressing here is just on the presentation layer.

My bad, I didn't notice you were also exposing the original request body, which makes sense, and thought I had to do some invasive changes.

Anyway, I hooked the existing button. I should probably implement something like `doesUrlRequireEncoding` and maybe rename some variables, but first let me know if something like this is OK.

vbuberen commented:
Joining the conversation here - agree with @cortinico that we should attach this feature to an existing button. However, we might need to reevaluate UI in next PRs (I suggest to do it in first of 4.x releases).

vbuberen commented on library/src/main/java/com/chuckerteam/chucker/internal/ui/transaction/TransactionPayloadFragment.kt:
```diff
@@ -71,21 +71,50 @@ internal class TransactionPayloadFragment :
-                    val result = processPayload(type, transaction)
-                    payloadBinding.responseRecyclerView.adapter = TransactionBodyAdapter(result)
-                    payloadBinding.responseRecyclerView.setHasFixedSize(true)
+                    val encode = viewModel.encodeUrl.value ?: true
```
Why don't you call this variable the same way as you named parameter `updateAdapter` fun?

gm-vm replied:
As I mentioned, I need to rename some stuff. I started with just the formatting of url encoded forms and ended up with more. I was wondering if I should stick with just the formatting of url encoded forms or continue with this more generic solution.

vbuberen commented on library/src/main/java/com/chuckerteam/chucker/internal/ui/transaction/TransactionPayloadFragment.kt:
```diff
@@ -71,21 +71,50 @@ internal class TransactionPayloadFragment :
+        viewModel.encodeUrl.observe(
+            viewLifecycleOwner,
+            Observer { encode ->
+                if (adapter.itemCount == 0) {
```
Why do you need this check if you also do the check for transaction value below?

gm-vm replied:
Definitely something I should have commented. There are cases in which `transaction` is not null here (e.g., configuration change), so you end up notifying the adapter twice: once here and once when observing `transaction`. Not a big deal, but it's unnecessary work.

I tried to think of a way not to have this here and the only thing I could think of is having a separate `transaction` LiveData that under the hood is just a MediatorLiveData that combines `transaction` and `encodeUrl`, taking care of notifying observers only once the first time it's observed. It didn't look pretty and it was more convoluted than this simple check, which is just a micro-optimization and not something required. With or without this check everything would work just fine. But I don't really like the fact that this trick only works if you start observing this LiveData before the other.

vbuberen commented on library/src/main/java/com/chuckerteam/chucker/internal/ui/transaction/TransactionPayloadFragment.kt:
```diff
@@ -113,7 +142,13 @@ internal class TransactionPayloadFragment :
-        menu.findItem(R.id.encode_url).isVisible = false
+        menu.findItem(R.id.encode_url).apply {
+            isVisible = true
+            setOnMenuItemClickListener {
```
This options item already had click listener in `TransactionActivity` and called exactly the same fun from ViewModel.

vbuberen commented on library/src/main/java/com/chuckerteam/chucker/internal/support/FormatUtils.kt:
```diff
@@ -94,6 +96,24 @@ internal object FormatUtils {
+                return form
+            }
+            form.split("&").joinToString(separator = "\n") { entry ->
+                val split = entry.split("=")
```
Rename this `val` please for more clarity.

vbuberen commented on library/src/main/java/com/chuckerteam/chucker/internal/ui/transaction/TransactionPayloadFragment.kt:
```diff
@@ -71,21 +71,50 @@ internal class TransactionPayloadFragment :
+        transaction: HttpTransaction,
+        formatted: Boolean
+    ) {
+        data.clear()
```
In general this function shouldn't be in this fragment. It should be inside the adapter itself named something like `setItems()`.

vbuberen replied:
Also, in case of having such thing as `setItems()` function you can remove adapter initialisation with an empty list and make its declaration lazy for cases when we have empty payload.

gm-vm replied:
Definitely, I didn't like this to begin with and I guess I just didn't want to change too many things. I'll fix this. But I'm not sure I get exactly what you meant with your proposed solution. Are you suggesting the use of a lazy property or something else?

Here I can only see a nullable list as marginally better, given that the moment you attach the adapter to the RecyclerView its `getItemCount()` gets called, so anything that creates the list lazily would do it immediately.

vbuberen replied:
Yes, I suggest to have a lazy property and create it only when we have some result to show.
However, now I think it might be not the best idea, since it will cause logcat show `No adapter attached`, like it does now.

gm-vm commented on library/src/main/java/com/chuckerteam/chucker/internal/ui/transaction/TransactionPayloadFragment.kt:
```diff
@@ -71,21 +71,50 @@ internal class TransactionPayloadFragment :
         )
     }
 
+    private suspend fun updateUpdater(
```
Definitely, I had a typo there and instead of fixing it my brain decided to make it worse.

gm-vm commented:
I rebased my changes and applied your suggestions.

I also changed the way things work. The button now only encodes/decodes url encoded requests.

I don't really like that `formatRequestBody` boolean there. I gave it a generic name, but we know that's a lie. I'd rather have the fragment not do anything and receive already formatted bodies. And maybe not as simple Strings, but something like a `Spannable`, which I see is what we end up with anyway. This would allow to format nicely certain type of requests.

For example, `application/x-www-form-urlencoded` requests are generally represented as I'm doing here, but keys and values are generally shown either with different colors or font weights. Then there are also `multipart/form-data`, where things get a bit more complicated.

vbuberen commented:
> I don't really like that `formatRequestBody` boolean there. I gave it a generic name, but we know that's a lie. I'd rather have the fragment not do anything and receive already formatted bodies.

Agree with you. It is something I also suggested recently. Look at `Next steps` here https://github.com/ChuckerTeam/chucker/pull/325#issue-403678329 . TL:DR - it left since times we had no view models. But it is going to change soon.

cortinico commented on library/src/main/java/com/chuckerteam/chucker/internal/data/entity/HttpTransaction.kt:
```diff
@@ -191,12 +191,16 @@ internal class HttpTransaction(
-            contentType != null && contentType.contains("xml", ignoreCase = true) ->
-                FormatUtils.formatXml(body)
-            else -> body
+        return if (contentType.isNullOrBlank()) {
+            body
+        } else {
+            when {
+                contentType.contains("json", ignoreCase = true) -> FormatUtils.formatJson(body)
+                contentType.contains("xml", ignoreCase = true) -> FormatUtils.formatXml(body)
+                contentType.contains("x-www-form-urlencoded", ignoreCase = true) ->
+                    FormatUtils.formatUrlEncodedForm(body)
+                else -> body
+            }
```
```suggestion
        return when {
            contentType.isNullOrBlank() -> body
            contentType.contains("json", ignoreCase = true) -> FormatUtils.formatJson(body)
            contentType.contains("xml", ignoreCase = true) -> FormatUtils.formatXml(body)
            contentType.contains("x-www-form-urlencoded", ignoreCase = true) ->
                FormatUtils.formatUrlEncodedForm(body)
            else -> body
        }
```

Please keep the single `return when` style as it's easier to read and follow.

cortinico commented on library/src/main/java/com/chuckerteam/chucker/internal/support/FormatUtils.kt:
```diff
@@ -94,6 +96,24 @@ internal object FormatUtils {
+                return form
+            }
+            form.split("&").joinToString(separator = "\n") { entry ->
+                val keyValue = entry.split("=")
+                val key = keyValue[0]
+                val value = if (keyValue.size > 1) URLDecoder.decode(keyValue[1], "UTF-8") else ""
+                "$key: $value"
```
```suggestion
entry.split("=").let { 
    (key, value) -> "$key: ${URLDecoder.decode(value, "UTF-8")}" 
}
```

You don't need extra variables as you can use a destructuring declaration.
Moreover you have to extend the catch below to include the `IndexOutOfBoundsException`

gm-vm replied:
My initial change had a destructuring declaration, but dropped it in favor of this to support what you can see in `testFormatUrlEncodedForm_noValues` (i.e., requests that need to send empty strings).

The current code should never throw an `IndexOutOfBoundsException`, since `split` always returns at least one string.

cortinico replied:
Makes sense, let's drop my comment. Thanks for clarifying.

cortinico commented on library/src/main/java/com/chuckerteam/chucker/internal/ui/transaction/TransactionViewModel.kt:
```diff
@@ -33,8 +33,23 @@ internal class TransactionViewModel(
+    val doesRequestBodyRequireEncoding: LiveData<Boolean> = RepositoryProvider.transaction()
+        .getTransaction(transactionId)
+        .map { transaction ->
+            if (transaction == null) {
+                false
+            } else {
+                transaction.requestContentType?.contains("x-www-form-urlencoded", ignoreCase = true) ?: false
+            }
```
```suggestion
            transaction?.requestContentType?.contains("x-www-form-urlencoded", ignoreCase = true) ?: false
```

vbuberen commented:
@cortinico So, we are good to merge now? I see that PR is updated already.

cortinico commented:
Yes 👍
